File: protocol-generator/dart/test/smoke_test.dart
import 'dart:typed_data';

import 'package:entity_core_protocol/entity_core_peer.dart';
import 'package:entity_core_protocol/src/codec/ecf_value.dart';
import 'package:test/test.dart';

/// S3 two-peer loopback smoke test (the phase exit gate).
///
/// Two Dart peers talk over real loopback TCP through the full §6.5 dispatch
/// chain. A RESPONDER peer listens; an INITIATOR peer (a second identity) dials
/// it and drives the §4.1 forward handshake (hello → authenticate), then:
///  - 404 on an unregistered path (no handler resolved);
///  - an authority-gated tree get (200) over the §4.4 discovery floor;
///  - a capability request (200);
///  - 8-way request_id demux of concurrently-issued replies (N7, §6.11).
///
/// A second scenario exercises the v7.74 Core Extensibility Boundary
/// (--debug-open-grants + --validate): the register live-hook (§6.13(a)), the
/// emit hook firing on register's tree writes (§6.13(c)), the §7a echo handler,
/// AND the §6.11 dispatch-outbound reentry (the validator-as-B-role surface S4's
/// origination-core needs).
///
/// The full validate-peer --profile core run is S4. This smoke proves the
/// wire-level peer surface so S4 can run the oracle.
void main() {
  final results = <bool>[];

  bool check(String name, bool ok) {
    results.add(ok);
    // ignore: avoid_print
    print('  [${ok ? "PASS" : "FAIL"}] $name');
    return ok;
  }

  Uint8List seed(int b) => Uint8List(32)..fillRange(0, 32, b);

  test('two-peer loopback (the S3 gate)', () async {
    await _runCoreScenario(check, seed);
    await _runExtensibilityScenario(check, seed);
    final pass = results.where((r) => r).length;
    final allPass = results.every((r) => r);
    // ignore: avoid_print
    print('\nSMOKE: ${allPass ? "PASS" : "FAIL"} ($pass/${results.length})');
    expect(allPass, isTrue, reason: 'two-peer loopback must be all-PASS');
  });
}

// ── Scenario 1: core ops (responder = default seed policy) ──────────────────────

Future<void> _runCoreScenario(
    bool Function(String, bool) check, Uint8List Function(int) seed) async {
  final responder = await Peer.create(seed(0x11));
  final listener = await startListener(responder, 0);
  try {
    // ignore: avoid_print
    print('Responder listening on 127.0.0.1:${listener.port} '
        '(peer ${responder.localPeer})');
    final initiator = await Peer.create(seed(0x22));
    final s = await dial(initiator, '127.0.0.1', listener.port);
    try {
      final remote = s.remotePeerId!;
      // ignore: avoid_print
      print('Handshake:');
      check('session established (capability minted)', s.capability != null);
      check('remote peer_id matches responder', remote == responder.localPeer);

      // ignore: avoid_print
      print('Dispatch:');
      final r404 =
          await s.execute('/$remote/does/not/exist', 'noop', emptyParams(), null);
      check('unregistered path -> 404', responseStatus(r404!) == 404);

      final ifaceTarget = resourceTarget(['system/handler/system/tree']);
      final rget =
          (await s.execute('/$remote/system/tree', 'get', emptyParams(), ifaceTarget))!;
      check('granted tree get -> 200', responseStatus(rget) == 200);
      final res = responseResult(rget);
      check('tree get returns a system/handler/interface entity',
          res != null && res.type == 'system/handler/interface');

      final reqGrant =
          Peer.grant(['system/tree'], ['system/type/*'], ['get'], null);
      final reqParams = Entity.make(
          'system/capability/request', cmap(['grants', EcfArray([reqGrant])]));
      final rcap =
          (await s.execute('/$remote/system/capability', 'request', reqParams, null))!;
      check('capability request -> 200', responseStatus(rcap) == 200);

      // 8-way request_id demux (N7, §6.11) — 8 concurrent futures.
      // ignore: avoid_print
      print('Concurrency (request_id demux):');
      final futures = [
        for (var i = 0; i < 8; i++)
          () async {
            final r = await s.execute('/$remote/system/tree', 'get',
                emptyParams(), resourceTarget(['system/handler/system/tree']));
            final rr = r == null ? null : responseResult(r);
            return r != null &&
                responseStatus(r) == 200 &&
                rr != null &&
                rr.type == 'system/handler/interface';
          }(),
      ];
      final correlated = (await Future.wait(futures)).where((b) => b).length;
      check('8 interleaved requests each correlated -> $correlated/8',
          correlated == 8);
    } finally {
      s.close();
    }
  } finally {
    await listener.close();
  }
}

// ── Scenario 2: the v7.74 Core Extensibility Boundary over the wire ─────────────

Future<void> _runExtensibilityScenario(
    bool Function(String, bool) check, Uint8List Function(int) seed) async {
  final responder =
      await Peer.create(seed(0x33), openGrants: true, conformance: true);
  var emitEvents = 0;
  responder.store.registerTreeConsumer((_) => emitEvents++);
  final listener = await startListener(responder, 0);
  try {
    final initiator =
        await Peer.create(seed(0x44), openGrants: true, conformance: true);
    final s = await dial(initiator, '127.0.0.1', listener.port);
    try {
      final remote = s.remotePeerId!;
      final emitBefore = emitEvents;
      // ignore: avoid_print
      print('Extensibility (open-grants + --validate):');

      // register live-hook (§6.13(a))
      final manifest = cmap(['name', 'demo', 'operations', emptyMap()]);
      final req =
          Entity.make('system/handler/register-request', cmap(['manifest', manifest]));
      final rreg = (await s.execute('/$remote/system/handler', 'register', req,
          resourceTarget(['system/handler/demo'])))!;
      check('handler register -> 200 (live, not 501)', responseStatus(rreg) == 200);
      check('emit hook fired on register tree writes (§6.13(c))',
          emitEvents > emitBefore);

      // §7a echo conformance handler (resolve→dispatch)
      final payload = Entity.make('primitive/any', cmap(['ping', EcfInt.of(42)]));
      final recho =
          (await s.execute('/$remote/system/validate/echo', 'echo', payload, null))!;
      check('§7a echo -> 200', responseStatus(recho) == 200);
      final res = responseResult(recho);
      check('§7a echo returns params verbatim',
          res != null && res.type == 'primitive/any');

      // §6.11 dispatch-outbound REENTRY (B→A echo over the inbound connection)
      check('§6.11 dispatch-outbound reentry round-trips (B→A echo over inbound)',
          await _runReentryProbe(s, remote));
    } finally {
      s.close();
    }
  } finally {
    await listener.close();
  }
}

/// Drive the §6.11 dispatch-outbound seam: the responder's (B) dispatch-outbound
/// handler originates an outbound EXECUTE back to the caller (A) over the SAME
/// inbound connection (reentry). The connection IS reentrant (B can write to A
/// over the open socket), so this proves the §6.13(b) outbound primitive + the
/// §6.11 reentry param shape end-to-end — the exact surface S4's origination-core
/// `dispatch_outbound_reentry` exercises over real two-peer TCP, here validated
/// at the smoke level (the validator supplies the cross-peer reentry cap at S4;
/// the smoke passes the session cap to prove the seam parses + reaches the
/// outbound primitive).
///
/// Accept outer 200 (A served the reentrant EXECUTE and B round-tripped it).
Future<bool> _runReentryProbe(Session s, String remote) async {
  final params = Entity.make(
      'primitive/any',
      cmap([
        'target', 'system/validate/echo',
        'operation', 'echo',
        'value', cmap(['ping', EcfInt.of(7)]),
        'reentry_capability', s.capability!.toCbor(),
        'reentry_granter', s.granterPeer!.toCbor(),
        'reentry_cap_signature', s.capSignature!.toCbor(),
      ]));
  final r = await s.execute(
      '/$remote/system/validate/dispatch-outbound', 'dispatch', params, null);
  if (r == null) return false;
  // Outer 200 = the §6.11 reentry round-tripped end-to-end: B's dispatch-outbound
  // handler originated an EXECUTE back to A over the SAME inbound connection, A's
  // reader dispatched it and replied, and B correlated the reply by request_id
  // and returned it. The INNER status reflects A's §5.2 authz verdict on the
  // reentrant EXECUTE (here the session cap A's granter B minted — a cap check,
  // not the transport). S4's validator supplies the cross-peer reentry cap that
  // makes the inner verdict 200; the smoke proves the transport seam (outer 200).
  final inner = responseResult(r)?.uint('status')?.toInt();
  // ignore: avoid_print
  print('    (reentry round-tripped; inner echo verdict=$inner)');
  return responseStatus(r) == 200;
}

File: protocol-generator/ruby/test/smoke_test.rb
# frozen_string_literal: true

require_relative "test_helper"

# S3 two-peer loopback smoke test (the phase exit gate).
#
# Two Ruby peers talk over real loopback TCP through the full §6.5 dispatch
# chain. A RESPONDER peer listens; an INITIATOR peer (a second identity) dials it
# and drives the §4.1 forward handshake (hello → authenticate), then:
#   * 404 on an unregistered path (no handler resolved);
#   * an authority-gated tree get (200) over the §4.4 discovery floor;
#   * a capability request (200);
#   * 8-way request_id demux of concurrently-issued replies (N7, §6.11).
# A second scenario exercises the v7.74 Core Extensibility Boundary
# (--debug-open-grants + --validate): the register live-hook (§6.13(a)), the emit
# hook firing on register's tree writes (§6.13(c)), and the §7a echo handler.
#
# The full validate-peer --profile core run is S4. This smoke proves the
# wire-level peer surface so S4 can run the oracle.
class SmokeTest < Minitest::Test
  include EntityCore

  def setup
    @results = []
  end

  def check(name, ok)
    @results << ok
    puts format("  [%<status>s] %<name>s", status: ok ? "PASS" : "FAIL", name: name)
    ok
  end

  def seed(byte)
    (byte.chr * 32).b
  end

  def test_two_peer_loopback
    run_core_scenario
    run_extensibility_scenario
    pass_count = @results.count(true)
    all_pass = @results.all?
    puts format("\nSMOKE: %<verdict>s (%<pass>d/%<total>d)",
                verdict: all_pass ? "PASS" : "FAIL", pass: pass_count, total: @results.size)
    assert all_pass, "two-peer loopback must be all-PASS (#{pass_count}/#{@results.size})"
  end

  # ── Scenario 1: core ops (responder = default seed policy) ──────────────────

  def run_core_scenario
    responder = Peer.create(seed(0x11))
    listener = Transport.start_listener(responder, 0)
    begin
      puts format("Responder listening on 127.0.0.1:%<port>d (peer %<peer>s)",
                  port: listener.port, peer: responder.local_peer)
      initiator = Peer.create(seed(0x22))
      session = Transport.dial(initiator, "127.0.0.1", listener.port)
      begin
        remote = session.remote_peer_id
        puts "Handshake:"
        check("session established (capability minted)", !session.capability.nil?)
        check("remote peer_id matches responder", remote == responder.local_peer)

        puts "Dispatch:"
        # 404 on an unregistered path
        r404 = session.execute("/#{remote}/does/not/exist", "noop", Wire.empty_params, nil)
        check("unregistered path -> 404", Wire.response_status(r404) == 404)

        # authority-gated tree get (200) over the discovery floor
        iface_target = Wire.resource_target("system/handler/system/tree")
        rget = session.execute("/#{remote}/system/tree", "get", Wire.empty_params, iface_target)
        check("granted tree get -> 200", Wire.response_status(rget) == 200)
        res = Wire.response_result(rget)
        check("tree get returns a system/handler/interface entity",
              !res.nil? && res.type == "system/handler/interface")

        # capability request (200)
        req_grant = Peer.grant(["system/tree"], ["system/type/*"], ["get"])
        req_params = Entity.make("system/capability/request", { "grants" => [req_grant] })
        rcap = session.execute("/#{remote}/system/capability", "request", req_params, nil)
        check("capability request -> 200", Wire.response_status(rcap) == 200)

        # 8-way request_id demux (N7, §6.11)
        puts "Concurrency (request_id demux):"
        n = 8
        oks = Array.new(n, false)
        threads = (0...n).map do |i|
          Thread.new do
            r = session.execute("/#{remote}/system/tree", "get", Wire.empty_params,
                                Wire.resource_target("system/handler/system/tree"))
            rr = Wire.response_result(r)
            oks[i] = Wire.response_status(r) == 200 && !rr.nil? && rr.type == "system/handler/interface"
          rescue StandardError
            oks[i] = false
          end
        end
        threads.each(&:join)
        correlated = oks.count(true)
        check("8 interleaved requests each correlated -> #{correlated}/8", correlated == n)
      ensure
        session.close
      end
    ensure
      listener.close
    end
  end

  # ── Scenario 2: the v7.74 Core Extensibility Boundary over the wire ─────────

  def run_extensibility_scenario
    responder = Peer.create(seed(0x33), open_grants: true, conformance: true)
    emit_events = 0
    emit_mutex = Mutex.new
    responder.store.register_tree_consumer { |_ev| emit_mutex.synchronize { emit_events += 1 } }
    listener = Transport.start_listener(responder, 0)
    begin
      initiator = Peer.create(seed(0x44))
      session = Transport.dial(initiator, "127.0.0.1", listener.port)
      begin
        remote = session.remote_peer_id
        emit_before = emit_mutex.synchronize { emit_events }
        puts "Extensibility (open-grants + --validate):"

        # register live-hook (§6.13(a))
        manifest = { "name" => "demo", "operations" => {} }
        req = Entity.make("system/handler/register-request", { "manifest" => manifest })
        rreg = session.execute("/#{remote}/system/handler", "register", req,
                              Wire.resource_target("system/handler/demo"))
        check("handler register -> 200 (live, not 501)", Wire.response_status(rreg) == 200)
        emit_after = emit_mutex.synchronize { emit_events }
        check("emit hook fired on register's tree writes (§6.13(c))", emit_after > emit_before)

        # §7a echo conformance handler (resolve→dispatch)
        payload = Entity.make("primitive/any", { "ping" => 42 })
        recho = session.execute("/#{remote}/system/validate/echo", "echo", payload, nil)
        check("§7a echo -> 200", Wire.response_status(recho) == 200)
        res = Wire.response_result(recho)
        check("§7a echo returns params verbatim", !res.nil? && res.type == "primitive/any")
      ensure
        session.close
      end
    ensure
      listener.close
    end
  end
end
